xurls.go

```diff
@@ -76,6 +76,23 @@ const (
 		`)`
 	ipAddr = `(?:` + ipv4Addr + `|` + ipv6Addr + `)`
 	port   = `(?::[0-9]*)?`
+
+	// authority is based on https://www.rfc-editor.org/rfc/rfc3987#section-2.2
+	// but with the same limitations as pathCont and a special exclusion of
+	// single-label domain names that are valid as an IPv6 address chomp
+	// to avoid creating output like `<a href="...">https://2001</a>:db8::1`
+	// from `2001:db8::1` while still matching names like `localhost`.
+	iUserinfo             = `[` + unreservedChar + `%` + subDelimChar + `:` + allowedUcsChar + `]*`
+	nameLabelSafeChar     = `a-zA-Z0-9\-_~`
+	iMidLabelChar         = nameLabelSafeChar + `%` + midSubDelimChar + allowedUcsChar
+	iEndLabelChar         = nameLabelSafeChar + `%` + endSubDelimChar + allowedUcsCharMinusPunc
+	iEndLabelCharMinusHex = `g-zG-Z\-_~%` + endSubDelimChar + allowedUcsCharMinusPunc
+	iRegNamePrefix        = `(?:[` + iMidLabelChar + `]{4,}[` + iEndLabelChar + `]\.?|` +
+		`[` + iMidLabelChar + `]{0,3}[` + iEndLabelCharMinusHex + `]\.?|` +
+		h4 + `\.)`
+	iRegName  = iRegNamePrefix + `(?:[` + iMidLabelChar + `]*[` + iEndLabelChar + `](?:\.[` + iMidLabelChar + `]*[` + iEndLabelChar + `])*\.?)?`
+	iHost     = `(?:\[` + ipv6Addr + `\]|` + ipv4Addr + `|` + iRegName + `)`
+	authority = `(?:` + iUserinfo + `@)?` + iHost + port
```

Comment:
Interesting; presumably this helps us match with more URLs, not less than before? If so, can we add more tests to lock in that behavior? I'm just having a slightly hard time reasoning around why we need all this extra regexp if the added test cases just cover trailing dots.

Or perhaps the alternative is to tweak `pathCont` so it doesn't match authorities. That might be best long-term, for the sake of not matching e.g. a `magnet` link with a port suffix.

Reply:
I'm not terribly worried about slower performance as long as the matching becomes better. A more exact regexp will almost always be longer, and that usually means it will be a bit slower.

Still, my comment above still applies - I'd like to better understand if the extra regexp is actually improving with current false negatives or false positives in terms of matching :)

Reply:
The motivation was an observation that "http://1080::8:800:200c:417a/path" was being incorrectly matched, and the strategy for fixing that was more obvious alignment with RFC 3987. I do think there are cases fixed by this other than IPv6 address bracketing and a final dot at the end of a domain name, though, but I'll have to mull over how to find them. At a glance, I note that `iRegName` is more limiting than `pathCont` (rejecting e.g. brackets, parentheses, braces, `\`, `:`, `@`, Private Use code points, and empty domain name labels `..`).

Reply:
Interestingly enough, GitHub has linkified that bad URL, and clicking on it seems to select the text rather than open the link :)